### PR TITLE
feat: support `ethabi` Contract methods

### PR DESCRIPTION
**Pull request by andyrobert3**
## Motivation

Fixes https://github.com/alloy-rs/core/issues/101

~Struggling to write tests for `load` method from `ethabi`~ 
Thanks @prestwich for the help!

## Solution

Reference from methods from `ethabi` 

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/json-abi/src/abi.rs [hunk 1]
```diff
@@ -8,6 +12,7 @@ use serde::{
     ser::SerializeSeq,
     Deserialize, Deserializer, Serialize,
 };
+use std::{io, iter::Flatten};
```
Comment (prestwich) [inline crates/json-abi/src/abi.rs:15]: `iter` should be used as `core::iter`. 

Any usage of `std` should be feature-wrapped

Comment (andyrobert3) [reply crates/json-abi/src/abi.rs:15]: How about `io`, can we still import from `std` @prestwich

**Comment by prestwich** [comment]
> value: Error("invalid type: string "address", expected a valid internal type", line: 1, column: 250)'

this seems odd. can you link to the ABI file that you're using for tests?

**Comment by prestwich** [comment]
okay i have tracked this down. 

- `InternalType` is deserialized by borrowing data from the deserializer with `Visitor::visit_borrowed_str`
- `serde_json::from_reader` does not call `Visitor::visit_borrowed_str`. 
- Instead it calls `Visitor::visit_string` which implicitly calls `Visitor::visit_str`. 
- Neither of those are implemented for the `ItVisitor`.
- Because the deserializer calls an unimplement `vist_` method, it returns the unexpected error, even tho the input is actually expected

**Comment by prestwich** [comment]
@andyrobert3 if you don't mind, I can push a fix to your branch. the fix will be loading the entire file into a string and then chaning to `read_string`, as well as making `serde_json::from_reader` always error

**Comment by andyrobert3** [comment]
sure @prestwich feel free to push!

need to sleep since its a bit late here

**Comment by prestwich** [comment]
cool i pushed. this solves the immediate issue, and should make it clearer for future users :)

File: crates/json-abi/src/abi.rs [hunk 2]
```diff
@@ -72,6 +79,55 @@ impl JsonAbi {
...
+        self.constructor.as_ref()
+    }
+
+    /// Loads contract from json
+    #[cfg(feature = "std")]
+    pub fn load<T: io::Read>(mut reader: T) -> Result<Self, serde_json::Error> {
+        // https://docs.rs/serde_json/latest/serde_json/fn.from_reader.html
+        // serde_json docs recommend buffering the whole reader to a string
+        // This also prevents a borrowing issue when deserializing from a reader
+        let mut json = String::with_capacity(1000);
```
Comment (DaniPopes) [inline crates/json-abi/src/abi.rs:94]: Power of two

```suggestion
        let mut json = String::with_capacity(1024);
```

**Review by DaniPopes — CHANGES_REQUESTED** [review]
Rest LGTM

File: crates/json-abi/src/abi.rs [hunk 3]
```diff
@@ -1,13 +1,20 @@
...
+use alloc::{
+    collections::{btree_map, btree_map::Values},
+    string::String,
+    vec::Vec,
+};
 use alloy_primitives::Bytes;
 use btree_map::BTreeMap;
-use core::{fmt, iter};
+use core::{fmt, iter, iter::Flatten};
+use serde_json as _;
```
[the remark this answers is not in the document]
Comment (prestwich) [reply crates/json-abi/src/abi.rs:10]: according to docs, `serde_json::from_str` does not require `std`, we can probably have

```
[features]
default = ["std", "serde_json"]
serde_json = ["dep:serde_json"]
```

the tests require both `"std"` and `"serde_json"`

[the remark this answers is not in the document]
Comment (prestwich) [reply crates/json-abi/src/abi.rs:10]: then the `load` method would be gated by `#[all(feature = "std", feature = "serde_json")]`

[the remark this answers is not in the document]
Comment (DaniPopes) [reply crates/json-abi/src/abi.rs:10]: right but then `serde_json` is unused when `"std"` is not enabled. I guess this is fine

**Comment by prestwich** [comment]
needs `cargo fmt`

File: crates/json-abi/src/abi.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (prestwich) [reply crates/json-abi/src/abi.rs:10]: that's pretty fair. are there any other `serde_json`-related convenience functions we want that don't necessarily require `std`?

i.e. should we have any interfaces like:
```
#[cfg(feature = "serde_json)]
pub fn from_json_string(s: impl AsRef<str>) -> Result<Self>;

$[cfg(all(feature = "serde_json", feature = "std"))]
pub fn to_json_string(&self) -> Result<String>;
```

[the remark this answers is not in the document]
Comment (prestwich) [reply crates/json-abi/src/abi.rs]: added this in a commit

**Comment by prestwich** [comment]
@DaniPopes mind re-reviewing with my additions? :)

**Review by DaniPopes — APPROVED** [review]

**Comment by andyrobert3** [comment]
Was going to work on the requested changes today, but thanks to @prestwich @DaniPopes for reviewing and making the required changes

**Comment by prestwich** [comment]
yeah, hope you don't mind us getting it over the finish line. thank you for your contribution! 🎉

**Pull request merged**